### PR TITLE
update `remove_enclosing_curly_braces`

### LINKED ISSUE #4: Add tests to `remove_enclosing_curly_braces` and use for qenv?
And simplify - as the function is not nice at all

From https://github.com/insightsengineering/coredev-tasks/issues/147

### PR DESCRIPTION
**Pull request by nikolas-burkoff**
Closes #4 

Compare this branch to main for existing module qenv's and also with for example:

```
library(teal.code)
library(teal.widgets)
library(shiny)


ui <- fluidPage(
  verbatim_popup_ui("err", "Error"),
  verbatim_popup_ui("ok", "OK")
)

server <- function(input, output, session) {

  q_ok <- reactive({
    new_qenv(
      list2env(list(y = 2, x = 1)),
      code = quote({
        x <- 1
        y <- 2
      })
    ) |> eval_code(quote({
      if (x == 1) {
        print("x = 1")
      }
    }))
  })


  verbatim_popup_srv(
    "ok",
    verbatim_content = reactive(get_code(q_ok())),
    title = "ok"
  )

  q_err <- reactive({
    new_qenv(
      list2env(list(y = 2)),
      code = quote({
        x <- 1
        y <- 2
      })
    ) |> eval_code(quote({
      if (x == 1) {
        print("x = 1")
      }
    }))
  })

  verbatim_popup_srv(
    "err",
    verbatim_content = reactive(get_code(q_err())),
    title = "Error"
  )

}

shinyApp(ui, server)

```
and

```
library(teal.code)
data_q <- new_qenv()
data_q <- eval_code(new_qenv(), "{iris_data <- iris}")
warning_qenv <- eval_code(
  data_q,
  bquote({p <- hist(iris_data[, .("Sepal.Length")], ff = "")})
)
cat(get_warnings(warning_qenv))
```

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -1,44 +1,41 @@
...
+      return(gsub(four_spaces_at_start_of_line, "", split_text))
     }
+  } else {
+    return(split_text)
   }
 }
+
+
+# converts vector of expressions to character
+format_expression <- function(code) {
```
Comment (nikolas-burkoff) [inline R/utils.R:39]: Note I have to leave `remove_enclosing_curly_braces` as a stand alone function as chunks use it.

File: R/utils.R [hunk 2]
```diff
@@ -1,44 +1,41 @@
-#' Removes curly brackets from string
+#' Removes leading and trailing curly brackets from character
+#' string and removes indentation of remaining contents
 #'
 #' @description `r lifecycle::badge("stable")`
 #' @param x (`character`)\cr
-#'  string with code chunk
 #'
-#' @return The string without curly braces
+#' @return Character string without curly braces
```
Comment (Polkas) [inline R/utils.R:7]:
```suggestion
#' @return `character(1)` string without curly braces.
```

File: R/utils.R [hunk 3]
```diff
@@ -1,44 +1,41 @@
...
     }
+  } else {
+    return(split_text)
   }
 }
+
+
+# converts vector of expressions to character
+format_expression <- function(code) {
+  unlist(lapply(as.character(code), remove_enclosing_curly_braces))
```
Comment (Polkas) [inline R/utils.R:40]: why not vapply with character(1)?:)

Comment (nikolas-burkoff) [reply R/utils.R:40]: image

Comment (nikolas-burkoff) [reply R/utils.R:40]: and I can't really change the output of `remove_enclosing_curly_braces` without changing chunks which I would rather not do...

Comment (nikolas-burkoff) [reply R/utils.R:40]: ...though I guess I could if we test to make sure chunks are still good

File: R/utils.R [hunk 4]
```diff
@@ -1,44 +1,41 @@
...
-    txt <- unlist(strsplit(x, "\n", fixed = TRUE))
-    if (grepl("^\\{[[:space:]]*", txt[1]) && grepl("^\\}[[:space:]]*", utils::tail(txt, 1))) {
-      txt2 <- txt[-c(1, length(txt))]
+    return(x)
+  }
+
+  open_bracket_then_spaces <- "^\\{[[:space:]]*"
+  close_bracket_then_spaces <- "^\\}[[:space:]]*"
+  blank_line <- "^[[:space:]]*$"
+  four_spaces_at_start_of_line <- "^[[:space:]]{4}"
```
Comment (Polkas) [inline R/utils.R:18]:
```suggestion
  open_bracket_then_spaces <- "^\\{[[:blank:]]*"
  close_bracket_then_spaces <- "^\\}[[:blank:]]*"
  blank_line <- "^[[:blank:]]*$"
  four_spaces_at_start_of_line <- "^[[:blank:]]{4}"
```

should be the same but more direct.

File: R/utils.R (hunk 2 above)
Comment (nikolas-burkoff) [reply R/utils.R]: so it doesn't always return character(1) - but I have removed the upper case "C"

File: R/utils.R (hunk 4 above)
Comment (nikolas-burkoff) [reply R/utils.R]: done

**Review by Polkas — APPROVED** [review]
LGTM, tested with sample apps and with code

```
library(teal.code)
a <- new_qenv(
  list2env(list(y = 2, x = 1)),
  code = quote({
    x <- 1
    y <- 2
  })
) |> eval_code(quote({
  if (x == 1) {
    print("x = 1")
  }
})) 

b <- new_qenv(
  list2env(list(y = 2, x = 1)),
  code = quote({
    x <- 1
    y <- 2
  })
) |> eval_code(quote({
  if (x == 1) {
    print("x = 1")
  }
}))

c <- a |> eval_code(quote({
  c <- 4
}))

concat(a, b)|> 
  get_code() |> paste(collapse = "\n") |> cat()

join(a, c)|> 
  get_code() |> paste(collapse = "\n") |> cat()


a2 <- new_qenv(
  list2env(list(y = 2, x = 1)),
  code = quote({
    {
      x <- 1
      y <- 2
    }
  })
) |> eval_code(quote({
  if (x == 1) print("x = 1")
})) 

a2 |> get_code() |> paste(collapse = "\n") |> cat()

a3 <- new_qenv(
  list2env(list(y = 2, x = 1)),
  code = quote({
    x <- 1; y <- 2
  })
) |> eval_code(quote({
  if (x == 1) print("x = 1"); print("Hello")
}))

a3 |> get_code() |> paste(collapse = "\n") |> cat()
```

we tested some edge cases too.

**Pull request merged**